### PR TITLE
migrations: Implement AdoptResources for the openstack provider

### PR DESCRIPTION
**Pull request by babbageclunk**
## Description of change
The AdoptResources environ method will be used at the end of a model migration to update the controller tag for all of the resources used by the model, so that if the source controller is subsequently destroyed then these resources won't be cleaned up with it.

In the openstack provider we tag volumes and instances with the controller UUID. Metadata for these can be updated using the API - I've added the SetVolumeMetadata method to the goose Cinder client (dependency updated here). Security groups don't have tags, so the controller is included as a component of the group name. The group names for the current model are all rewritten with the new controller UUID.

## QA steps
None yet - this method isn't called from the migration master yet.

## Documentation changes
N/A

## Bug reference
This is part of the fix for https://bugs.launchpad.net/juju/+bug/1648063

### CODE REVIEW

File: provider/openstack/provider.go [hunk 1]
```diff
@@ -1261,7 +1261,53 @@ func (e *Environ) Instances(ids []instance.Id) ([]instance.Instance, error) {
...
+
+	instances, err := e.AllInstances()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cinder, err := e.cinderProvider()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// I'm not sure about this - it works at the moment because
```
Comment (babbageclunk) [inline provider/openstack/provider.go]: Oops - I meant to pull this out and replace it with a comment on github. Anyway, I'll pull it out once I know the answer.

File: provider/openstack/firewaller.go [hunk 2]
```diff
@@ -21,6 +21,11 @@ import (
 	"github.com/juju/juju/network"
 )
 
+var (
+	validUUID                = `[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}`
+	ExtractGroupControllerRe = regexp.MustCompile(`^(?P<prefix>juju-)(?P<controllerUUID>` + validUUID + `)(?P<suffix>-.*)$`)
```
Comment (axw) [inline provider/openstack/firewaller.go]: Please don't export the regexp itself. The pattern would be OK exported as a constant, but this thing is mutable.

File: provider/openstack/firewaller.go [hunk 3]
```diff
@@ -49,6 +54,10 @@ type Firewaller interface {
 	// DeleteGroups deletes the security groups with the specified names.
 	DeleteGroups(names ...string) error
 
+	// Implementations are expected to update all of the security
```
Comment (axw) [inline provider/openstack/firewaller.go]: I realise the function below doesn't follow the pattern, but please stick to the idiom of:

"UpdateGroupController updates all of the security groups for this model to refer to the specified controller."

I think the intent might be a bit clearer by appending something like ", such that DeleteAllControllerGroups will remove them only when called with the specified controller ID."

File: provider/openstack/provider.go (hunk 1 above)
Comment (axw) [inline provider/openstack/provider.go]: maybe just replace this with "TODO(axw) fix the storage API" :)

File: provider/openstack/firewaller.go (hunk 2 above)
Comment (babbageclunk) [inline provider/openstack/firewaller.go]: Ah, ok - I'll export the string as a const.

File: provider/openstack/firewaller.go (hunk 3 above)
Comment (babbageclunk) [inline provider/openstack/firewaller.go]: Sorry, will fix. Yeah, adding the rationale will make it clearer.

File: provider/openstack/provider.go (hunk 1 above)
Comment (babbageclunk) [inline provider/openstack/provider.go]: OK :)

**Comment by babbageclunk** [comment]
!!chittychitty!!

**Comment by babbageclunk** [comment]
$$merge$$

**Pull request merged**